relay-server/src/services/project_cache.rs

```diff
@@ -899,6 +865,42 @@ impl ProjectCacheBroker {
         }
     }
 
+    /// Returns `true` if the project state valid for the [`QueueKey`].
+    ///
+    /// Which includes the own key and the samplig key for the project.
+    fn is_state_valid(&mut self, key: &QueueKey) -> bool {
+        let QueueKey {
+            own_key,
+            sampling_key,
+        } = key;
+
+        let is_own_state_valid = self.projects.get_mut(own_key).map_or(false, |project| {
+            // Returns `Some` if the project is cache otherwise None and also triggers refresh
+            // in background.
+            project
+                .get_cached_state(self.services.project_cache.clone(), false)
+                // Makes sure that the state also is valid.
+                .map_or(false, |state| !state.invalid())
+        });
```

Comment:
not really significant but an early return here if own_state isn't valid should be a bit more performant

Reply:
This is done on purpose, since I still want to trigger update for the `sampling_key` if it's there and also expired, so next iteration will be much faster